### PR TITLE
Fix `DropdownMenu` focus

### LINKED ISSUE #143505: DropdownMenu doesn't fire onSelected from keyboard navigation
STEPS TO REPRODUCE

0. Run the program below.
1. Tab until the dropdown menu is focused. (Side note: this seems to take an inordinate number of tabs given that there's only two focusable controls on the screen, but that's [another bug](https://github.com/flutter/flutter/issues/131120).)
2. Press enter to open the menu.
3. Hit down arrow to highlight "Foo.aa".
4. Press enter to select "Foo.aa".

EXPECTED RESULTS

Selected value becomes "Foo.aa".

ACTUAL RESULTS

Selected value remains "null".

CODE FOR TEST PROGRAM

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MaterialApp(home: MyHomePage()));
}

enum Foo { aa, bb, cc }

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  Foo? _value;

  @override
  Widget build(BuildContext context) {
    return Material(
      child: Column(
        children: [
          Text('Selected value: $_value'),
          DropdownMenu<Foo>(
            onSelected: (Foo? value) {
              setState(() {
                _value = value;
              });
            },
            dropdownMenuEntries: Foo.values
                .map<DropdownMenuEntry<Foo>>(
                  (Foo f) => DropdownMenuEntry(value: f, label: '$f'),
                )
                .toList(),
          ),
        ],
      ),
    );
  }
}
```

### PR DESCRIPTION
**Pull request by yiiim**
Fixes #143505

I am unable to test sending `TextInputClient.performAction` when the enter key is pressed on desktop platforms. I have created an issue for this: #156414. The current test approach is to check whether the `DropdownMenu` has focus.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

### CODE REVIEW

File: packages/flutter/lib/src/material/dropdown_menu.dart [hunk 1]
```diff
@@ -506,6 +507,7 @@ class _DropdownMenuState<T> extends State<DropdownMenu<T>> {
   @override
   void initState() {
     super.initState();
+    _focusNode = widget.focusNode ?? FocusNode();
```
Comment (navaronbracke) [inline packages/flutter/lib/src/material/dropdown_menu.dart:510]: We should follow the pattern where we have:
a) A user-provided FocusNode
b) An internally managed FocusNode, if the user never provided one

So you might want to have an "_effectiveFocudeNode" getter for the other parts of this change.

File: packages/flutter/lib/src/material/dropdown_menu.dart [hunk 2]
```diff
@@ -490,6 +490,7 @@ class _DropdownMenuState<T> extends State<DropdownMenu<T>> {
   double? leadingPadding;
   bool _menuHasEnabledItem = false;
   TextEditingController? _localTextEditingController;
+  FocusNode? _focusNode;
```
Comment (navaronbracke) [inline packages/flutter/lib/src/material/dropdown_menu.dart:493]: When this focus node is created internally, it should also be disposed by this widget

File: packages/flutter/lib/src/material/dropdown_menu.dart (hunk 1 above)
Comment (yiiim) [reply packages/flutter/lib/src/material/dropdown_menu.dart]: done

File: packages/flutter/test/material/dropdown_menu_test.dart [hunk 3]
```diff
@@ -3452,6 +3452,42 @@ void main() {
...
+    /*Now the focus is on the icon button.*/
+
+    await tester.sendKeyEvent(LogicalKeyboardKey.enter);
+    await tester.pump();
+
+    await tester.sendKeyEvent(LogicalKeyboardKey.arrowDown);
+    await tester.pump();
+
+    expect(focusNode.hasPrimaryFocus, true);
+  }, variant:TargetPlatformVariant.desktop());
```
Comment (bleroux) [inline packages/flutter/test/material/dropdown_menu_test.dart:3490]:
```suggestion
  });
```

Keyboard navigation is also possible on mobile.
After removing this variant, set `requestFocusOnTap` to true on the DropdownMenu (line 3466), this way the initial focus state will be the same for mobile and desktop.
See https://github.com/flutter/flutter/pull/156131 where the desktop variant was removed for several tests.

File: packages/flutter/test/material/dropdown_menu_test.dart [hunk 4]
```diff
@@ -3452,6 +3452,42 @@ void main() {
...
+
+    await tester.sendKeyEvent(LogicalKeyboardKey.tab);
+    await tester.pump();
+
+    await tester.sendKeyEvent(LogicalKeyboardKey.tab);
+    await tester.pump();
+
+    await tester.sendKeyEvent(LogicalKeyboardKey.tab);
+    await tester.pump();
+    /*Now the focus is on the icon button.*/
```
Comment (bleroux) [inline packages/flutter/test/material/dropdown_menu_test.dart:3481]:
```suggestion

    // Now the focus is on the icon button.
    final Element iconButton = tester.firstElement(find.byIcon(Icons.arrow_drop_down));
    expect(Focus.of(iconButton).hasFocus, isTrue);
```

**Review by bleroux — COMMENTED** [review]
Interesting fix, thanks! 🙏 

Some minor comments on the test side.

**Comment by justinmc** [comment]
@yiiim FYI there is a merge conflict.

**Review by bleroux — COMMENTED** [review]
@yiiim I tried to run the two tests on master (without your change) and they did not failed.
Maybe this was fixed by a recent change (several DropdownMenu PRs landed in the past weeks)?

**Review by QuncCccccc — COMMENTED** [review]
This looks good to me! Thanks for your contribution!

> I am unable to test sending TextInputClient.performAction when the enter key is pressed on desktop platforms.

As for the test, can we try `await tester.testTextInput.receiveAction(TextInputAction.done);`?

**Comment by QuncCccccc** [comment]
> @yiiim I tried to run the two tests on master (without your change) and they did not failed. Maybe this was fixed by a recent change (several DropdownMenu PRs landed in the past weeks)?

The problem wasn't fixed by the recent landed PRs. I think this PR fix makes sense to me:)!

**Comment by yiiim** [comment]
> This looks good to me! Thanks for your contribution!
> 
> > I am unable to test sending TextInputClient.performAction when the enter key is pressed on desktop platforms.
> 
> As for the test, can we try `await tester.testTextInput.receiveAction(TextInputAction.done);`?

The current testing method still has issues. `Focus.of(editableText).hasFocus` cannot accurately determine whether the `TextField` has focus, and `await tester.testTextInput.receiveAction(TextInputAction.done);` cannot test this scenario either, as it always triggers `onEditingComplete` regardless of where the focus is. I plan to use `tester.state<EditableTextState>(find.byType(EditableText)).wantKeepAlive` to determine whether the `TextField` has focus. However, I am not ready with all the code yet, and I will update this PR soon.

**Comment by yiiim** [comment]
I changed the fix approach. We can't give focus to the `TextField` because, by default, the `TextField` is read-only on mobile platforms and cannot receive focus. If we continue to do so, keyboard navigation will still not work on mobile platforms. I had to use a less elegant solution by hiding an extra `Focus` widget to handle keyboard navigation. I tried giving focus directly to the MenuItem, but it breaks the highlight. Do you have any other ideas?

The good news is that I was able to test it successfully.

**Review by QuncCccccc — COMMENTED** [review]
Thanks so much for continuing to work on the fix! Overall this looks good to me! Just have one comment below. I found the problem when I did a manual test by using the sample code provided in the issue.

Could you also help fix the conflict? Sorry for letting this PR sit here for such a long time!

File: packages/flutter/test/material/dropdown_menu_test.dart [hunk 5]
```diff
@@ -3576,6 +3574,93 @@ void main() {
       expect(tester.getRect(findMenuMaterial()).top, textFieldBottom);
     });
   });
+  // Regression test for https://github.com/flutter/flutter/issues/143505.
```
Comment (bleroux) [inline packages/flutter/test/material/dropdown_menu_test.dart:3577]:
```suggestion
  });
  
  // Regression test for https://github.com/flutter/flutter/issues/143505.
```

File: packages/flutter/lib/src/material/dropdown_menu.dart [hunk 6]
```diff
@@ -1041,8 +1048,28 @@ class _DropdownMenuState<T> extends State<DropdownMenu<T>> {
...
           onInvoke: handleDownKeyInvoke,
         ),
+        _EnterIntent: CallbackAction<_EnterIntent>(
+          onInvoke: (_) => _handleEditingComplete(),
+        ),
       },
-      child: menuAnchor,
+      child: Stack(
+        children: <Widget>[
+          // Handling keyboard navigation when the Textfield has no focus
```
Comment (bleroux) [inline packages/flutter/lib/src/material/dropdown_menu.dart:1057]:
```suggestion
          // Handling keyboard navigation when the Textfield has no focus.
```

**Review by bleroux — APPROVED** [review]
LGTM! with two very minor nits.

File: packages/flutter/test/material/dropdown_menu_test.dart [hunk 7]
```diff
@@ -3576,6 +3574,94 @@ void main() {
       expect(tester.getRect(findMenuMaterial()).top, textFieldBottom);
     });
   });
+
```
Comment (bleroux) [inline packages/flutter/test/material/dropdown_menu_test.dart:3577]:
```suggestion
 
```

Sorry for the trailing whitespace which breaks "Linux Analyse" check.

**Review by QuncCccccc — APPROVED** [review]
LGTM:) Thanks a lot for the contribution!

**Pull request merged**